Add support for opening trigger modal via URL pre-populated

viiccwen opened the pull request:
## Description

This PR introduces a new feature that allows **calling DAG trigger modals** with **Single/Backfill** mode  to be opened directly via URL using the query string parameter.

The feature also supports **pre-populating form fields** from URL parameters, improving automation and integration with external tools.

## Changes

### 1. Trigger Modal Opening/Closing
* When visiting a DAG page with `/dags/{dag_id}/trigger/single` or `/dags/{dag_id}/trigger/backfill`, the Trigger Modal automatically opens with correspond mode.
* When closing the trigger modal, the pathname will turn to `/dags/{dag_id}` automatically.

### 2. URL Parameter Pre-Population

Trigger form supports (you can check its functionality on UI):
* `conf`
* `run_id`
* `logical_date`
* `note`

Backfill form supports (you can check its functionality on UI):
* `start_date`
* `end_date`
* `run_backwards`
* `reprocess_behavior`
* `max_active_runs`
* `conf`

`conf` supports 2 type:

| Form                 | Usage                                                                                                         | Example                                                                                     |
| -------------------- | ------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------- |
| JSON (explicit)      | Provide the entire configuration as a JSON object.<br>Has higher priority if present.                         | `/dags/{dag_id}/trigger/single?conf={"foo":"bar","x":123}`                                  |
| Key-value (implicit) | If `conf` is not specified, any query parameter that is not a reserved keyword will be collected into `conf`. | `/dags/{dag_id}/trigger/single?run_id=myrun&foo=bar&x=123` → `conf={"foo":"bar","x":"123"}` |

### 3. Utilities Added

New helper file in `src/utils/trigger.ts`:
1. put preloaded function (`getPreloadTriggerFormData`, `getPreloadBackfillFormData`)
2. `getRunModeFromPathname`
3. normalize the trigger path by params `normalizeTriggerPath`, i.e. (`/trigger/single`, `/trigger/backfill`)

### 4. Advanced Options
* The **Advanced Options** section is expanded automatically when pre-filled values are detected from URL parameters.

## Examples

### 1. Single mode
`/trigger/single&mode=single?conf={"environment":"production","debug":true,"user":"admin"}&run_id=test_run_001&logical_date=2024-01-15T10:30:00.000&note=testing note`
### 2. Backfill mode
`/trigger/backfill?start_date=2024-01-01&end_date=2024-01-31&run_backwards=true&reprocess_behavior=missing_runs&max_active_runs=3&conf={"environment":"staging","batch_size":1000}`

## Snapshot

1. single mode with `/dag/{dag_id}/trigger/single?run_id=run%20dag&logical_date=2025-09-06T12:34:56.789&conf={%22foo%22:%22bar%22,%22ping%22:%22pong%22}&note=Nothing%20beats%20a%20jet2%20holiday`


https://github.com/user-attachments/assets/d299309e-3425-4a2f-a055-506fbd007bd9



2. backfill mode with `/dags/{dag_id}/trigger/backfill?start_date=2025-09-01T12:34:56&end_date=2025-09-02T12:34:56&conf={%22foo%22:%22bar%22,%22ping%22:%22pong%22}&reprocess_behavior=completed&max_active_runs=3&run_backwards=1`


https://github.com/user-attachments/assets/f94eaa6e-6c02-46fd-8bf6-6d47af277167



3. `key=value` type of conf with `/dags/{dag_id}/trigger/single?run_id=run%20dag&logical_date=2025-09-06T12:34:56.789&foo=bar&ping=pong&note=Nothing%20beats%20a%20jet2%20holiday`


https://github.com/user-attachments/assets/5011b785-1b3b-4f01-b671-c28bf048412f



## Documentation

I edited the `/core-concepts/params.html` file for this change, I added `trigger params`, `conf usage table`, `examples for 2 mode`.


## Related Issue

resolves #54800

viiccwen committed:
feat(ui): allow TriggerDAGButton to auto-open via URL param

Added support for `?trigger=true` or `?trigger=1` query parameter to automatically open TriggerDAGModal

viiccwen committed:
feat(utils): add URL param decoding and safe JSON parsing helpers

- Added `decodeParam` for safe URL decoding
- Added `parseJsonSafe` for safe JSON.parse with fallback
- Added `getUrlParam` to unify parameter extraction and optional JSON parsing

viiccwen committed:
feat(ui): support URL param pre-population for TriggerDAGForm

- Pre-populate fields (conf, dag_run_id, logical_date, note) from URL parameters
- URL conf is validated and pretty-printed if valid JSON
- add openAdvanced that supports open automatically

viiccwen committed:
feat: add URL param for choosing single or backfill mode

viiccwen committed:
feat: support URL param pre-population for BackfillForm

- Added parsing of backfill-specific parameters (start_date, end_date, reset_dagruns, rerun_failed_tasks, etc.)
- Connected URL param values into Backfill form default state
- add mapped helper function with ReprocessBehavior

viiccwen committed:
fix: URL config param filled functionality

viiccwen committed:
feat: enhancement for removing the params when closing the modal

jscheffl commented on airflow-core/src/airflow/ui/src/components/TriggerDag/TriggerDAGButton.tsx:
```diff
@@ -31,9 +33,20 @@ type Props = {
+  const { search } = useLocation();
+  const params = useMemo(() => new URLSearchParams(search), [search]);
+
+  useEffect(() => {
```
The usage of `useEffect()` is restricted and should only be used when there is no other alternative.

jscheffl commented on airflow-core/src/airflow/ui/src/constants/reprocessBehaviourParams.ts:
```diff
@@ -16,8 +16,23 @@
+    case "missing_runs":
+      return "none";
+    default:
+      return "none";
```
Default can be used for the missing as well
```suggestion
    default:
      return "none";
```

jscheffl commented on airflow-core/src/airflow/ui/src/constants/reprocessBehaviourParams.ts:
```diff
@@ -16,8 +16,23 @@
   { label: "backfill.allRuns", value: "completed" },
 ];
+
+export const mapReprocessBehavior = (param: string): ReprocessBehavior => {
```
Why do you map the strings from the API to the user URL? Would it not be easier to use the same in the URL as well? Just probably needs some documentation so that users know about it.

jscheffl commented on airflow-core/src/airflow/ui/src/utils/query.ts:
```diff
@@ -47,3 +47,45 @@ export const useAutoRefresh = ({ dagId, isPaused }: { dagId?: string; isPaused?:
+    return JSON.stringify(parsed, null, 2);
+  } catch {
+    // Return original string if parsing fails
+    return raw;
```
Why do you return the original if parsing fails? Then it would not be "safe" and some follow-up code will fail on this?

jscheffl commented on airflow-core/src/airflow/ui/src/utils/query.ts:
```diff
@@ -47,3 +47,45 @@ export const useAutoRefresh = ({ dagId, isPaused }: { dagId?: string; isPaused?:
+    return decodeURIComponent(value);
+  } catch {
+    // Return original value if decoding fails
+    return value;
```
Same as below - why returning the original string if decoding fails? This would allow passing "bad" data further into later code and makes it very hard to understand bugs?

jscheffl reviewed (commented):
Thanks for the PR. Some comments as general code review. Did not test myself (yet) as I assume it will need a bit of adjustments. Code is mainly looking good.

Regarding your questions:
- I (personally, other opinions welcome) would favor using `/trigger` and not a URL parameter for this. Also the `mode` could be directly encoded as part of the URL like `/trigger/single` (and would propose to have single as default because this would be for 90% of calls I assume).

General feedback to the PR:
- In PR #54783 we also had a discussion about passing form values as JSON dict. In my view this makes it complex because a user who constructs a URL must make a proper JSON and encode the data in the URL properly. Finally we came to the point in passing individual key/vale pairs and not a JSON dict. See also https://github.com/apache/airflow/pull/54783#issuecomment-3217154737 - I'd wish to make it consistent here as well.
- Please also add documentation back to airflow-core/docs/core-concepts/params.rst as it was removed in https://github.com/apache/airflow/pull/54799 and describe how users can use it (would be bad if users need to read source to discover such cool features!)

viiccwen committed:
refactor: remove redundant param `missing_runs` to return `none`

viiccwen committed:
refactor: remove redundant function

viiccwen committed:
feat: change trigger and mode from params to pathname

1. add `/trigger/single` and `/trigger/backfill` pathname to fit the correct mode.
2. add trigger.ts: encapsulate params getter function to new file
3. refactor TriggerDAGModal, TrggerDAGForm, TrggerDAGButton, RunBackfillForm

* RunBackfillForm is almost closed to the maximum line, we may need to refactor it.

viiccwen committed:
docs(params): add trigger params and trigger example

viiccwen committed:
feat: supports conf `key=value` key in method if not provided explicit `conf` params

viiccwen committed:
docs(params): add `conf` params usage table between JSON and `key=value``

guan404ming commented on airflow-core/docs/core-concepts/params.rst:
```diff
@@ -349,10 +349,61 @@ The following features are supported in the Trigger UI Form:
+  - ``conf`` – JSON configuration.
+  - ``run_id`` – run identifier.
+  - ``logical_date`` – execution date in ``YYYY-MM-DDTHH:mm:ss.SSS`` format. Defaults to the current timestamp if not provided.
+  - ``note`` – note attached to the DAG run.
```
The fields in the form would dynamically change based on user's config. Thus, we may not only need to handle these specific fields but we should handle all fields passed. The implementation is quite like https://github.com/apache/airflow/pull/54783

guan404ming reviewed (commented):
Thanks for the PR!

I think the implementation could be simpler by just get search param via hook and preload them in `FlexiableForm`

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/DagActions/RunBackfillForm.tsx:
```diff
@@ -54,15 +57,26 @@ const RunBackfillForm = ({ dag, onClose }: RunBackfillFormProps) => {
   const { conf } = useParamStore();
-  const { control, handleSubmit, reset, watch } = useForm<BackfillFormProps>({
+  const { search } = useLocation();
+  const params = new URLSearchParams(search);
+  const {
+    conf: urlConf,
+    from_date: urlFromDate,
+    max_active_runs: urlMaxActiveRuns,
+    reprocess_behavior: urlReprocessBehavior,
+    run_backwards: urlRunBackwards,
+    to_date: urlToDate,
+  } = getPreloadBackfillFormData(params);
```
We should use the `searchParam` hook there.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/components/TriggerDag/TriggerDAGModal.tsx:
```diff
@@ -67,6 +73,19 @@ const TriggerDAGModal: React.FC<TriggerDAGModalProps> = ({
   const nameOverflowing = dagDisplayName.length > maxDisplayLength;
 
+  useEffect(() => {
+    // Only sync URL when already within trigger route to avoid interfering with close navigation
+    if (!open || !/\/trigger(\/(single|backfill))?$/u.test(pathname)) {
+      return;
+    }
+
+    const targetPath = normalizeTriggerPath(pathname, runMode);
+
+    if (pathname !== targetPath) {
+      navigate({ pathname: targetPath, search }, { replace: true });
+    }
+  }, [runMode, pathname, search, navigate, open]);
+
```
You can probably add router entries on `..../trigger/{mode}` that also maps to the same page, and handle the path param `mode` if in the URL.

pierrejeambrun commented on airflow-core/src/airflow/ui/src/utils/trigger.ts:
```diff
@@ -0,0 +1,81 @@
+
+export const getPreloadBackfillFormData = (searchParams: URLSearchParams) => ({
+  conf: getPreloadTriggerConf(searchParams, ["max_active_runs", "reprocess_behavior", "run_backwards"]),
+  from_date: searchParams.get("start_date") ?? "",
+  max_active_runs: parseInt(searchParams.get("max_active_runs") ?? "1", 10),
+  reprocess_behavior: (searchParams.get("reprocess_behavior") ?? "none") as ReprocessBehavior,
+  run_backwards: ["1", "true"].includes(searchParams.get("run_backwards") ?? ""),
+  to_date: searchParams.get("end_date") ?? "",
```
why do you change the `url` name of the param vs the one used for the form?
```suggestion
  from_date: searchParams.get("from_date") ?? "",
  max_active_runs: parseInt(searchParams.get("max_active_runs") ?? "1", 10),
  reprocess_behavior: (searchParams.get("reprocess_behavior") ?? "none") as ReprocessBehavior,
  run_backwards: ["1", "true"].includes(searchParams.get("run_backwards") ?? ""),
  to_date: searchParams.get("to_date") ?? "",
```

Also you should probably use the PARAMS constants from `SearchParamsKeys` and update that enum if some keys are missing.

pierrejeambrun reviewed (commented):
Nice, a few suggestion in addition to @guan404ming comments.

bbovenzi commented on airflow-core/src/airflow/ui/src/router.tsx:
```diff
@@ -173,6 +173,15 @@ export const routerConfig = [
+      {
+        children: [
+          { element: <Overview />, index: true },
+          { element: <Overview />, path: "single" },
+          { element: <Overview />, path: "backfill" },
```
I wonder if we should pass `trigger={"single" | "backfill" | undefined}` as props from here instead of trying to manually parse the url string?

bbovenzi commented on airflow-core/src/airflow/ui/src/components/TriggerDag/TriggerDAGModal.tsx:
```diff
@@ -79,7 +98,7 @@ const TriggerDAGModal: React.FC<TriggerDAGModalProps> = ({
         </Dialog.Header>
 
-        <Dialog.CloseTrigger />
+        <Dialog.CloseTrigger onClick={onClose} />
```
We don't need this change

viiccwen committed:
Merge branch 'main' into feature/ui-trigger-form-url-params
